File: my_code/DSA/JS/DS/linked-lists/singly-linked-list/middle-of-linked-list [Slow, fast pointers].js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
  }

  // pushing to the end of the list O(n)
  push(val) {
    // Receive a value and make a new node
    const newNode = new Node(val);

    // if there's no head => empty list
    // so set the head to be the new Node
    if (!this.head) {
      this.head = newNode;
      return this;
    }

    // if not empty, loop until the end
    let current = this.head;
    while (current.next) {
      current = current.next;
    }

    // then make the last node point into the new node
    current.next = newNode;
    return this;
  }

  middleNode() {
    // pointer move one step each time
    let step = this.head;
    // pointer move two steps each time
    let twoSteps = this.head;

    // while the second pointer not null, and has another node after it
    while (twoSteps && twoSteps.next) {
      // move one step
      step = step.next;
      // move two steps
      twoSteps = twoSteps.next.next;
    }

    // so when the second pointer reaches the end, the first pointer will reach the middle
    // after stop the loop will return the middle node
    return step;
  }

  display() {
    // print the linked list
    let current = this.head;
    while (current) {
      process.stdout.write(current.value + " ");
      current = current.next;
    }
    console.log("None");
  }
}

// Example usage
const myList = new LinkedList();
for (let i = 1; i <= 5; i++) {
  myList.push(i);
}

myList.display();
const mid = myList.middleNode();
console.log("Middle node value is:", mid.value);

// Using Two Pointers pattern
// Slow, fast pointers
// THIS SOLUTION TAKE Time => O(n), and Space O(1)
// BEST PRACTICE

File: my_code/DSA/PYTHON/DS/linked lists/singly-linked-list/middle-of-linked-list [Slow, fast pointers].py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    # pushing to the end of the list O(n)
    def push(self, val):
        #  Receive a value and make a new node
        new_node = Node(val)

        # if there's no head => empty list
        # so set the head to be the new Node
        if not self.head:
            self.head = new_node
            return self

        # if not empty, loop until the end
        current = self.head
        while current.next:
            current = current.next

        # then make the last node point into the new node
        current.next = new_node
        return self

    def middle_node(self):
        # pointer move one step each time
        step = self.head
        # pointer move two steps each time
        two_steps = self.head

        # while the second pointer not none, and has another node after it
        while two_steps and two_steps.next:
            # move one step
            step = step.next
            # move two step
            two_steps = two_steps.next.next
        
        # so when the second pointer reach the end, the first pointer will reach the middle
        # after stop the loop will return the middle node
        return step

    def display(self):
        # print the linked list
        current = self.head
        while current:
            print(current.value, end=" ")
            current = current.next
        print("None")


my_list = LinkedList()
for i in range(1, 6):  
    my_list.push(i)

my_list.display()
mid = my_list.middle_node()
print("Middle node value is:", mid.value)

# Using Two Pointers pattern
# Slow, fast pointers
# THIS SOLUTION TAKE Time => O(n), and Space O(1)
# BEST PRACTICE